✨Feat: 결제관련 예외처리 추가

lazzzykim opened the pull request:
## 의도
내가 쓴 게시물에 결제할 수 없습니다.
내가 쓴 게시물에 리뷰를 작성할 수 없습니다.
판매가 완료된 상품은 isSoldOut 값이 true로 변경됩니다.
주문상태를 관리할 수 있습니다.
결제 취소시 환불받을 수 있습니다.
판매 완료된 상품은 구매할 수 없습니다.

- 내용

## 주요 변경점
어떤 부분이 변경되었는지 적어주세요
- refundToAccount 메서드를 추가하여 환불 받을 수 있는 코드가 추가되었습니다.
- OrderState 클래스를 추가해 결제상태를 COMPLED, CANCELLED 로 관리할 수 있습니다.
- 주문 생성 시 주문상태가 COMPLED로 변경됩니다. 취소 시 CANCELLED로 변경됩니다.
- 내가 쓴 게시물에 결제할 시 IllegarStateException이 발생합니다.
- 내가 쓴 게시물에 리뷰를 작성할 시 IllegarStateException이 발생합니다. (이 부분은 결제한 사람만 리뷰를 작성하도록 수정할 계획입니다.)
- 판매 완료된 상품에 결제를 시도할 시 IllegarStateException이 발생합니다.

## ETC
리뷰어들에게 어떤 점을 검토해야하는지 알려주세요
- OrderServiceImpl의 createOrder부분이 수정되었으니 이 부분을 중점으로 검토 부탁드립니다.

lazzzykim committed:
✨Feat: 결제 취소 및 환불기능 구현

Account: OrderServiceImpl에서 환불할 금액을 처리하기 위해 refundToAccount 메서드 생성
Order: 주문 상태를 변경하기 위해 orderState EnumType 컬럼 추가
OrderController: 주문취소 요청을 처리하는 cancelOrder 메서드 추가
OrderService: 주문을 취소하는 함수 cancelOrder 추가
OrderServiceImpl:
createOrder 글 작성 시 주문상태를 COMPLED로 설정하도록 변경,
cancelOrder 주문 취소, 계좌에 환불,주문상태를 CANCELLED로 변경
OrderState: 주문 상태를 CANCELLED, COMPLED 로 관리

lazzzykim committed:
✨Feat: 자신의 글에 리뷰를 작성할 수 없는 기능 추가

Post: Entity에 내 게시글에 리뷰를 작성할 수 없게 하기위해 myPostCheckPermission 메서드 추가
ReviewServiceImpl: createReview 메서드에 myPostCheckPermission 메서드 적용

resolves: #6

lazzzykim committed:
Merge branch 'feature/order' into feature/review

# Conflicts:
#	src/main/kotlin/com/teamsparta/buysell/domain/post/model/Post.kt

lazzzykim committed:
✨Feat: 자신의 글에 리뷰를 작성할 수 없는 기능 추가

Post: Entity에 내 게시글에 리뷰를 작성할 수 없게 하기위해 myPostCheckPermission 메서드 추가

resolves: #6

lazzzykim committed:
🚜Fix: Account Entity에 SoftDeleteEntity import추가

Account: SoftEntity가 import되지 않은 버전을 push해서 re push

resolves: #8

lazzzykim committed:
✨Feat: 결제 관련 예외 추가

Post:
outOfStockStatus 게시글이 품절인지 확인하고, 품절인 경우 예외를 발생하는 메서드 추가
myPostCheckPermission 리뷰와 결제 코드에서 사용하기 위해 에러 메세지 수정

resolves: #8

lazzzykim committed:
✨Feat: 리뷰 softDelete 쿼리 추가

Review: @SQLDelete 추가, edit 메서드 명 변경 (edit -> editReview)
ReviewServiceImpl: 메서드명 수정 review.edit(request) -> review.editReview(request)

resolves: #6

lazzzykim committed:
Merge remote-tracking branch 'origin/dev' into feature/order

HwangSeungHyeon commented on src/main/kotlin/com/teamsparta/buysell/domain/member/model/Account.kt:
```diff
@@ -29,4 +29,7 @@ class Account (
     }
-}
\ No newline at end of file
+    fun refundToAccount(money: Long) {
```
refundToAccount와 depositToAccount가 함수 명을 제외하면 완벽하게 동일한데, 코드 중복을 없애는 것도 나쁘지 않을 것 같습니다.

lazzzykim replied:
코드 리팩토링 때 수정하도록 하겠습니다!

HwangSeungHyeon commented on src/main/kotlin/com/teamsparta/buysell/domain/order/model/Order.kt:
```diff
@@ -18,10 +18,14 @@ class Order (
 
     @OneToOne(fetch = FetchType.LAZY )
-    var post: Post
+    var post: Post,
```
혹시 member는 val을 쓰셨는데, post는 var를 쓰신 이유가 있으신가요?

lazzzykim replied:
var 보단 val이 맞는 것 같습니다. 수정했습니다!

codekmj1 reviewed (approved):
확인 했습니다

HwangSeungHyeon commented on src/main/kotlin/com/teamsparta/buysell/domain/order/service/OrderServiceImpl.kt:
```diff
@@ -29,17 +30,45 @@ class OrderServiceImpl(
             member = member,
-            post = post
+            post = post,
+            orderState = OrderState.COMPLED
```
COMPLED 대신 완료를 뜻하는 COMPLETED 같은 걸 사용해보는 건 어떠신가요?

lazzzykim replied:
번역을 쓰다보니..ㅎㅎ 수정했습니다

lazzzykim committed:
🔨Refactor: OrderState 클래스 수정

COMPLED -> COMPLETED로 수정
Order: var post: Post -> val post: Post 수정

resolves: #8

HwangSeungHyeon reviewed (approved):
고생 많으셨습니다!
